tests/e2e/tests/checkout/checkout.page.ts

```diff
@@ -0,0 +1,252 @@
+		const postcode = billingForm.locator( '#billing-postcode' );
+
+		await email.fill( customerBillingDetails.email );
+		await firstName.fill( customerBillingDetails.firstname );
+		await lastName.fill( customerBillingDetails.lastname );
+		await country.fill( customerBillingDetails.country );
+		await address1.fill( customerBillingDetails.addressfirstline );
+		await address2.fill( customerBillingDetails.addresssecondline );
+		await city.fill( customerBillingDetails.city );
+		await phone.fill( customerBillingDetails.phone );
+
+		if ( await state.isVisible() ) {
+			await state.fill( customerBillingDetails.state );
+		}
+		if ( await postcode.isVisible() ) {
+			await postcode.fill( customerBillingDetails.postcode );
+		}
+		// Blur active field to trigger customer address update, then wait for requests to finish.
+		await this.page.evaluate( 'document.activeElement.blur()' );
+		await this.checkCustomerPushCompleted(
+			'billing',
+			customerBillingDetails
+		);
+	}
+
+	async fillShippingDetails( customerShippingDetails ) {
+		const shippingForm = this.page.getByRole( 'group', {
+			name: 'Shipping address',
+		} );
+		const companyInputField = shippingForm.getByLabel( 'Company' );
+
+		if ( await companyInputField.isVisible() ) {
+			await companyInputField.fill( customerShippingDetails.company );
+		}
+
+		const firstName = shippingForm.getByLabel( 'First name' );
+		const lastName = shippingForm.getByLabel( 'Last name' );
+		const country = shippingForm.getByLabel( 'Country/Region' );
+		const address1 = shippingForm.getByLabel( 'Address', { exact: true } );
+		const address2 = shippingForm.getByLabel( 'Apartment, suite, etc.' );
+		const city = shippingForm.getByLabel( 'City' );
+		const state = shippingForm.getByLabel( 'State', { exact: true } );
+		const phone = shippingForm.getByLabel( 'Phone' );
+
+		// Using locator here since the label of this form changes depending on the country.
+		const postcode = shippingForm.locator( '#shipping-postcode' );
+
+		await firstName.fill( customerShippingDetails.firstname );
+		await lastName.fill( customerShippingDetails.lastname );
+		await country.fill( customerShippingDetails.country );
+		await address1.fill( customerShippingDetails.addressfirstline );
+		await address2.fill( customerShippingDetails.addresssecondline );
+		await city.fill( customerShippingDetails.city );
+		await phone.fill( customerShippingDetails.phone );
+
+		if ( await state.isVisible() ) {
+			await state.fill( customerShippingDetails.state );
+		}
+		if ( await postcode.isVisible() ) {
+			await postcode.fill( customerShippingDetails.postcode );
+		}
+		// Blur active field to trigger customer address update, then wait for requests to finish.
+		await this.page.evaluate( 'document.activeElement.blur()' );
+		await this.checkCustomerPushCompleted(
+			'shipping',
+			customerShippingDetails
+		);
+	}
+
+	async checkCustomerPushCompleted(
```

Comment:
How much value is there in this test vs comparing the address on the confirmation page? If the push didn't work, the data would be wrong right?

Reply:
The confirmation page will only show if the user is logged in, if they're not the addresses won't show.

The address is also output as a single block of text, broken up with `<br>` elements and linebreaks, so writing a function to verify the address is present is a little tricky. I just spent some time trying various things out and you end up with something like this:

```js
let addressString = `${ testData.firstname } ${ testData.lastname }<br>${ testData.addressfirstline }<br>`;

		if ( testData.addresssecondline ) {
			addressString += `${ testData.addresssecondline }<br>`;
		}
		if ( testData.city ) {
			addressString += `${ testData.city }<br>`;
		}
		if ( testData.state ) {
			addressString += `${ testData.state }<br>`;
		}
		if ( testData.postcode ) {
			addressString += `${ testData.postcode }<br>`;
		}
```

Because this function, `fillInCheckoutWithTestData` is supposed to be generic, and can be overridden with additional data, I think we're better off verifying the push, rather than the order confirmation.

I think we can have a focused test where we verify the order received page shows the correct address, where we use a specific address and know exactly what to expect.

Reply:
Noting that the Jest test suite tests both ^ and that guest users should see the address if they have a valid URL? I don't think its hidden in the current version 🤔

Reply:
I made some changes. Now the user is Iogged in, they select a shipping method and check out. The addresses are verified on the order confirmation page. Next, they select a different shipping method and enter a different billing address. Then the different addresses are verified on the order confirmation page. It's better to favour longer running tests as we cut down on setup and teardown time. These tests make sense to be grouped together, too.